### ISSUE TITLE
[MEDIUM] `react-native-keyboard-controller` migration plan

### ISSUE THREAD

**Issue by kirillzyusko** [Expensify/App#37203]
## Problem statement

In Expensify app keyboard handling is not consistent across iOS/Android platforms:
- on Android we mostly rely on default keyboard avoiding mechanism `adjustResize` (it works great, but lacks of synchronous animations - which means the screen will be instantly resized, and there will be no animation at all. For chat-like apps this is quite important to have for good UX)
- since Android handles it automatically - it's very hard to add any customization to keyboard handling (for example to preserve elements position which are obscured by keyboard)
- on iOS we are using JS code mixture to handle keyboard avoidance (`KeyboardAvoidingView`, subscriptions to keyboard events)
since iOS keyboard handling is done in JS code, on Android we use native keyboard avoidance provided by OS and we are writing cross-platform code -> we have a lot of Platform dependent code (like enabling `KeyboardAvoidingView` only on iOS), which eventually leads to UI inconsistency
- react-native offers a limited API for keyboard handling (especially on Android, where we can detect only `Did` events)
- react-native API is far behind from modern ways of keyboard handling

With `react-native-keyboard-controller`:
- we are getting cross-platform behavior (done in https://github.com/Expensify/App/pull/52392):
     - edge-to-edge mode -> identical work with insets on both platforms
     - disabled default keyboard handling - all keyboard handling is delegated to developer providing basics primitives
- consistent API for working with the keyboard (all events are available on both platforms, including `Will` events which were not available prior to RN API)
- interactive keyboard support - an ability to dismiss keyboard via gesture (default RN implementation only works on iOS and has severe limitations, such as `multiline` TextInput can not grow)

## Migration plan

I see next potential way for improving keyboard handling (aka migration plan):
- ~~implement inline auto suggestions (https://github.com/Expensify/App/pull/42630)~~
- ~~start to use it on a chat screen on iOS only (https://github.com/Expensify/App/pull/54764)~~
- ~~start to use this library on chat screen on Android (this library also will give a smooth transition on chat screen when keyboard appears/disappears, so UI will look better & more pleasant for end user) (https://github.com/Expensify/App/pull/54101)~~
- ~~search screen:~~
   - ~~right now "footer" (invite a friend, get 250$) has an instant transition and doesn't follow keyboard movement - with this library transition will be smooth (`KeyboardStickyFooter` component) (https://github.com/Expensify/App/pull/54101)~~ 

   <details><summary>See it in action</summary><img width="457" alt="image" src="https://github.com/Expensify/App/assets/22820318/d63eacf1-17f5-4492-a7ce-c665e79c2e64"></details>

   - ~~on Android when keyboard appears -> at the bottom of the screen we have a different color (under keyboard) - that's how default keyboard avoidance works in Android - when we'll have a frame-level control we can fix it as well and we'll not see any blinks there (https://github.com/Expensify/App/pull/54101)~~

   <details><summary>See it in action</summary><img width="459" alt="image" src="https://github.com/Expensify/App/assets/22820318/b0ecee05-24e0-472e-ba69-5b8178653c6b"></details>

- on Chat screen we can add "interactive keyboard dismissal" (when keyboard can be closed via [gesture](https://kirillzyusko.github.io/react-native-keyboard-controller/docs/guides/interactive-keyboard)) - https://github.com/Expensify/App/issues/54270
- @roryabraham  wanted to replace `useKeyboardContext` with this new library
- In Profile->Address screen: auto scroll is waiting for `keyboardDidShow` event and we have kind of two-fold animation -> potentially can be fixed by [KeyboardAwareScrollView](https://kirillzyusko.github.io/react-native-keyboard-controller/docs/api/components/keyboard-aware-scroll-view) with fully synchronized animations 

## Additional resources
- Slack discussion: https://expensify.slack.com/archives/C01GTK53T8Q/p1707467001447909
- GH issue of library review: https://github.com/Expensify/App/issues/35270
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01360efa253e7a2c60</li>
        <li>Upwork Job ID: 1762155306344951808</li>
        <li>Last Price Increase: 2024-04-23</li>
    </ul>
</details>

**Comment by roryabraham** [comment]
this is kind of halfway between a new feature and a bug, but I'm going to treat it as a bug for the sake of accounting (no CAP SW project it fits with)

**Comment by roryabraham** [comment]
Marking this as `Internal`, because I don't want to create an external Upwork job, but I _do_ want a C+ to help review all the PR(s) for this. I think it will be a series of PRs, so I think we should compensate for the review of each.

**Comment by roryabraham** [comment]
welcome @cubuspl42! I encourage you to familiarize yourself with the context on this issue and decide whether you want to take it on or pass it off. Excited to have you help if you're interested!

**Comment by roryabraham** [comment]
moving this to weekly

**Comment by roryabraham** [comment]
Fair enough. This is a broader migration than just that issue/PR though, basically we'll be refactoring keyboard handling across the whole app.

**Comment by cubuspl42** [comment]
Sounds interesting; I can get on board. This will be implemented by an expert agency, and C+ reviewed, right?

**Comment by cubuspl42** [comment]
@shubham1206agra I took a look. It's a PR with a 1-year-long post history. Let me know what exact aspect of that PR you'dl like me to focus on!

**Comment by roryabraham** [comment]
> This will be implemented by an expert agency, and C+ reviewed, right?

correct. 2 C+ in this case

**Comment by roryabraham** [comment]
@kirillzyusko looks like we need to resolve conflicts in https://github.com/Expensify/App/pull/16356, then @shubham1206agra and @cubuspl42 can begin review whenever they're ready

**Comment by roryabraham** [comment]
Also, this is not a bug

**Comment by JmillsExpensify** [comment]
Still working on this one.

**Comment by cubuspl42** [comment]
Yes, we found some bugs but they are already fixed. I'll give it another testing round today, I also bumped @shubham1206agra so they give it another pair of eyes.

**Comment by roryabraham** [comment]
agree, we'll pick this back up when @kirillzyusko is back from OOO. He works for Margelo, btw 🙂

**Comment by mallenexpensify** [comment]
@shubham1206agra reassigning, please take over as C+. If you don't have bandwidth, unassign yourself. Thanks

**Comment by mallenexpensify** [comment]
@rojiphil reassigning, please take over as C+. If you don't have bandwidth, unassign yourself. Thanks

**Comment by roryabraham** [comment]
@kirillzyusko any thoughts on when you'll be prioritizing this?

**Comment by kirillzyusko** [comment]
@roryabraham I think the current plan is to finish `native-stack` integration and after that I'll switch back to this task 👀 

In a meantime @perunt is also making some progress on integrating `react-native-keyboard-controller` into Expensify codebase (recently I added a cursor position tracking with coordinates and it will give us an ability to avoid RN core patching and push a task with emoji autosuggestion forward).

**Comment by JmillsExpensify** [comment]
Are we still working on finishing `native-stack` ahead of this task, or is work resuming elsewhere via another issue?

**Comment by kirillzyusko** [comment]
@JmillsExpensify I'm working on this issue (and I updated issue description to mention that one PR link has been changed).

**Comment by roryabraham** [comment]
@kirillzyusko what's the plan here? Should we keep this open or revisit later?.

**Comment by kirillzyusko** [comment]
@roryabraham we are actively working on that!

@perunt already started to use the functionality of this library in https://github.com/Expensify/App/pull/42630 and now he is actively working on https://github.com/Expensify/App/pull/42143

Meanwhile I'm working on https://github.com/Expensify/App/pull/47096 (this PR is kind of a preparation before https://github.com/Expensify/App/pull/42143)

When we start to use more functionality of `react-native-keyboard-controller` we can try spread the usage of this library across the codebase and start to use it on Android etc.

So I'd like to keep it open since the progress is moving!

**Comment by roryabraham** [comment]
Looks like https://github.com/Expensify/App/pull/42630 and https://github.com/Expensify/App/pull/47096 were merged, and https://github.com/Expensify/App/pull/42143 is still WIP

**Comment by perunt** [comment]
It's in the [review phase](https://github.com/Expensify/App/pull/42143#issuecomment-2353780455). Once testers confirm there's no jittering on their devices, we'll be good to go.
I asked about such testing because I was only able to test it on a few devices. On my oldest, least reliable device (an iPhone 7 with significant throttling and overheating issues), it shows jittering. However, the rest of the devices, such as the iPhone XR and 15, work well. So, the jittering might not be representative of typical user experiences

**Comment by roryabraham** [comment]
Need someone to hand this off to. @luacmartins not sure who's best, but going to reassign to you. Maybe you can handle or pass off to another Expensify engineer

**Comment by luacmartins** [comment]
@perunt are we still waiting for a review on https://github.com/Expensify/App/pull/42143?

**Comment by kirillzyusko** [comment]
@luacmartins yes, review + we need to fix issues that were found 👀

**Comment by luacmartins** [comment]
Still working on the PR.

**Comment by luacmartins** [comment]
Working on v2 of the PR since the first one got reverted

**Comment by JmillsExpensify** [comment]
Still working on the PRs

**Comment by JmillsExpensify** [comment]
Same same

**Comment by JmillsExpensify** [comment]
In progress

**Comment by luacmartins** [comment]
WIP

**Comment by roryabraham** [comment]
Resuming my stewardship of this issue. Seems like it overlaps substantially with @hannojg's proposal [here](https://expensify.slack.com/archives/C05LX9D6E07/p1734434242143269)

**Comment by mallenexpensify** [comment]
From @hannojg 's post [here](https://expensify.slack.com/archives/C05LX9D6E07/p1739874026164349?thread_ts=1734434242.143269&cid=C05LX9D6E07) last week
> We are currently working on finishing this PR, which brings react native keyboard controller to the Report/Chat screen:
- https://github.com/Expensify/App/pull/54764
Once thats fixed we will continue with working on this task, which will improve the keyboard interaction on that screen:
- https://github.com/Expensify/App/issues/54270

@roryabraham is this issue holding on the PR or issue linked above?  (It's HIGH on #quality and I'm reviewing all the HIGH issues cuz I'll be doing updates (likely bi-weekly for HIGH ones) in #quality soon.

**Comment by roryabraham** [comment]
This task is indeed on HOLD on https://github.com/Expensify/App/pull/54764 to the best of my knowledge, but that is an ancient PR and I'm not sure it's being treated with the priority of HIGH on #quality. For some more context, the [first incarnation of this PR](https://github.com/Expensify/App/pull/16356) was opened nearly two years ago: in March, 2023.

**Comment by roryabraham** [comment]
I think this can be put on HOLD for https://github.com/Expensify/App/issues/10632

**Comment by JmillsExpensify** [comment]
Awesome, looking forward to it!

**Comment by roryabraham** [comment]
Not overdue - PR is moving

**Comment by JmillsExpensify** [comment]
Sure looks like it.

**Comment by JmillsExpensify** [comment]
Still on hold

**Comment by roryabraham** [comment]
@kirillzyusko I think we can move forward with this one?

**Comment by kirillzyusko** [comment]
Yes @roryabraham 

@wgsquayson is currently working on https://github.com/Expensify/App/pull/65517 to bring interactive dismissal to Expensify.

The last pieces will be `useKeyboardContext` replacement (I think it can be achieved with `useKeyboardState` from `react-native-keyboard-controller`) and [KeyboardAwareScrollView](https://kirillzyusko.github.io/react-native-keyboard-controller/docs/api/components/keyboard-aware-scroll-view) usage on some screens (such as Profile) 👀

**Comment by JmillsExpensify** [comment]
What's the latest on this issue?

**Comment by roryabraham** [comment]
development continues in two distinct PRs: https://github.com/Expensify/App/pull/65517#issuecomment-3344103998

**Comment by roryabraham** [comment]
Work now continues in https://github.com/Expensify/App/pull/71438

**Comment by roryabraham** [comment]
I think we have done everything we set out to do here.
